iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/compaction/selector/impl/SettleSelectorImpl.java

```diff
@@ -0,0 +1,328 @@
+    } catch (Exception e) {
+      LOGGER.error(
+          "{}-{} cannot select file for settle compaction", storageGroupName, dataRegionId, e);
+    }
+    return Collections.emptyList();
+  }
+
+  private DirtyStatus selectFileBaseOnModSize(TsFileResource resource) {
+    ModificationFile modFile = resource.getModFile();
+    if (modFile == null || !modFile.exists()) {
+      return DirtyStatus.NOT_SATISFIED;
+    }
+    return modFile.getSize() > config.getInnerCompactionTaskSelectionModsFileThreshold()
+            || (!heavySelect
+                && !CompactionUtils.isDiskHasSpace(
+                    config.getInnerCompactionTaskSelectionDiskRedundancy()))
+        ? PARTIAL_DELETED
+        : DirtyStatus.NOT_SATISFIED;
+  }
+
+  /**
+   * Only when all devices with ttl are deleted may they be selected. On the basic of the previous,
+   * only when the number of deleted devices exceeds the threshold or has expired for too long will
+   * they be selected.
+   *
+   * @return dirty status means the status of current resource.
+   */
+  private DirtyStatus selectFileBaseOnDirtyData(TsFileResource resource)
+      throws IOException, IllegalPathException {
+    ModificationFile modFile = resource.getModFile();
+    DeviceTimeIndex deviceTimeIndex =
+        resource.getTimeIndexType() == ITimeIndex.FILE_TIME_INDEX_TYPE
+            ? resource.buildDeviceTimeIndex()
+            : (DeviceTimeIndex) resource.getTimeIndex();
+    Set<IDeviceID> deletedDevices = new HashSet<>();
+    boolean hasExpiredTooLong = false;
+    long currentTime = CommonDateTimeUtils.currentTime();
+
+    Collection<Modification> modifications = modFile.getModifications();
+    for (IDeviceID device : deviceTimeIndex.getDevices()) {
+      // check expired device by ttl
+      long deviceTTL = DataNodeTTLCache.getInstance().getTTL(((PlainDeviceID) device).toStringID());
+      boolean hasSetTTL = deviceTTL != Long.MAX_VALUE;
+      boolean isDeleted =
+          !deviceTimeIndex.isDeviceAlive(device, deviceTTL)
+              || isDeviceDeletedByMods(
+                  modifications,
+                  device,
+                  deviceTimeIndex.getStartTime(device),
+                  deviceTimeIndex.getEndTime(device));
+
+      if (hasSetTTL) {
+        if (!isDeleted) {
+          // For devices with TTL set, all data must expire in order to meet the conditions for
+          // being selected.
+          return DirtyStatus.NOT_SATISFIED;
+        }
+        long outdatedTimeDiff = currentTime - deviceTimeIndex.getEndTime(device);
+        hasExpiredTooLong =
+            hasExpiredTooLong
+                || outdatedTimeDiff > Math.min(config.getMaxExpiredTime(), 3 * deviceTTL);
+      }
+
+      if (isDeleted) {
+        deletedDevices.add(device);
+      }
+    }
+
+    float deletedDeviceRate = (float) (deletedDevices.size()) / deviceTimeIndex.getDevices().size();
+    if (deletedDeviceRate == 1f) {
```

Comment:
Is this comparison of float numbers valid?

Reply:
Resolved. Convert `int` to `double` type for division to avoid precision loss.